jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,486 @@
+     * <p>Priority {@code 0} is the lowest priority.</p>
+     * <p>The set of returned priorities should be stable over
+     * time, typically constrained in the range {@code 0-10}.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    /**
+     * <p>Fails the given suspended request/response with the given error code and failure.</p>
+     * <p>This method is called only for suspended requests, in case of timeout while suspended,
+     * or in case of failure when trying to handle a resumed request.</p>
+     *
+     * @param request the request to fail
+     * @param response the response to fail
+     * @param callback the callback to complete
+     * @param status the failure status code
+     * @param failure the failure
+     */
+    protected void failSuspended(Request request, Response response, Callback callback, int status, Throwable failure)
+    {
+        Response.writeError(request, response, callback, status, null, failure);
+    }
+
+    private boolean forward(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} forwarding {}", this, request);
+        request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
+        return super.handle(request, response, callback);
+    }
+
+    private void suspend(Request request, Response response, Callback callback)
+    {
+        int priority = Math.max(0, getPriority(request));
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} suspending priority={} {}", this, priority, request);
+        Entry entry = new Entry(request, response, callback, priority);
+        queues.compute(priority, (k, v) ->
+        {
+            if (v != null)
+                return v;
+            priorities.add(priority);
+            return new ConcurrentLinkedQueue<>();
+        }).offer(entry);
+        timeouts.schedule(entry);
+    }
+
+    private void resume()
+    {
+        // Only increment the permits if there are no suspended requests.
+        // In this way, new requests will not steal the permits,
+        // so that already suspended requests will not starve.
+        while (true)
+        {
+            if (tryResumeSuspended())
+                return;
+
+            // Here there are no suspended requests, so increment the permits.
+            // See correspondent code in handle() and expire().
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (state.compareAndSet(encoded, permits + 1, 0))
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} no suspended requests to resume", this);
+                return;
+            }
```

Comment:
again this hard coded zero itches me when you have not actually checked that permits are zero.  You are only inferring it because you didn't find a suspended request and thus assume encoded will change if the suspended requests change.   Far better to be explicit:
```suggestion
            int permits = AtomicBiInteger.getHi(encoded);
            int suspended = AtomicBiInteger.getLo(encoded);
            if (suspended > 0)
                continue; // spin and look for a suspended request again
            if (state.compareAndSet(encoded, permits + 1, suspended))
            {
                if (LOG.isDebugEnabled())
                    LOG.debug("{} no suspended requests to resume", this);
                return;
            }
```